app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb

```diff
@@ -42,21 +42,94 @@ def get_uuid_from_target(object)
+    }.freeze
+
     def refresh
       ems.with_provider_connection(VERSION_HASH) do |connection|
         @connection = connection
-        res = {}
-        res[:cluster] = collect_clusters
-        res[:storage] = collect_storages
-        res[:host] = collect_hosts
-        res[:vm] = collect_vms
-        res[:template] = collect_templates
-        res[:network] = collect_networks
-        res[:datacenter] = collect_datacenters
+        top_level_fututres_hash = collect_top_level_futures
+        res = wait_on_top_level_collections(top_level_fututres_hash)
+        res.each { |inv_name, inv| res[inv_name] = collect_inv_with_attributes_async(inv, inv_name_to_inv_type(inv_name)) }
+        preloaded_disks = collect_disks_as_hash(res[:disk])
+        collect_disks_from_attachments(res[:vm], preloaded_disks)
+        collect_disks_from_attachments(res[:template], preloaded_disks)
         res
       end
     end
 
+    def collect_disks_from_attachments(inv, preloaded_disks)
+      inv.each do |inv_item|
+        disks = AttachedDisksFetcher.attached_disks(inv_item, connection, preloaded_disks)
+        inv_item.singleton_class.send(:attr_accessor, :disks)
+        inv_item.disks = disks
+      end
+    end
+
+    def collect_inv_with_attributes_async(inv, inv_type)
+      return inv if ATTRIBUTES_TO_FETCH_FOR_INV_TYPE[inv_type].blank?
+      inv_keyed_requests = collect_inv_attributes_request(inv, inv_type)
+      inv_fetched_attributes = ManageIQ::Providers::Redhat::InfraManager::FuturesCollector.process_keyed_requests_queue(inv_keyed_requests)
+      raise "Failed to fetch attributes for #{inv_type}" unless inv_fetched_attributes
+      set_inv_attributes!(inv, inv_fetched_attributes, inv_type)
+    end
+
+    def apply_hosts_fetched_results!(hosts, fetched_host_results)
+      set_inv_attributes!(hosts, fetched_host_results, :host)
+    end
+
+    def set_inv_attributes!(inv, inv_fetched_attributes, inv_type)
+      inv.each do |obj|
+        key = base_key_for_obj(obj, inv_type)
+        ATTRIBUTES_TO_FETCH_FOR_INV_TYPE[inv_type].each do |attribute_name|
+          obj.instance_variable_set("@#{attribute_name}", inv_fetched_attributes["#{key}#{attribute_name}"])
+        end
+      end
+    end
+
+    def collect_inv_attributes_request(objs, obj_type)
+      obj_requests = []
+      objs.each do |obj|
+        ATTRIBUTES_TO_FETCH_FOR_INV_TYPE[obj_type].each do |attribute_name|
+          obj_requests << obj_attribute_request(obj, obj_type, attribute_name)
+        end
+      end
+      obj_requests
+    end
+
+    def wait_on_top_level_collections(top_level_fututres_hash)
+      top_level_fututres_hash.each do |k, v|
+        top_level_fututres_hash[k] = v.wait
+      end
+      top_level_fututres_hash
+    end
+
+    def collect_top_level_futures
+      res = {}
+      %w(cluster storage host vm template network datacenter disk).each { |inv_name| res[inv_name.to_sym] = collect_inv_future(inv_name.to_sym) }
```

Comment:
pls extract `%w(cluster storage host vm template network datacenter disk)` to constant

Reply:
done